File: sem4/mpp/ex1/networking/src/main/java/networking/worker/AbstractJsonWorker.java
package networking.worker;

import com.google.gson.Gson;
import networking.Request;
import networking.Response;

import java.io.IOException;
import java.net.Socket;

public abstract class AbstractJsonWorker extends AbstractWorker {
	private final Gson gsonFormatter;

	public AbstractJsonWorker(Socket connection) {
		super(connection);
		gsonFormatter = new Gson();
	}

	protected Request readRequest() throws IOException {
		String requestLine = input.readLine();
		System.out.println("request received " + requestLine);
		return gsonFormatter.fromJson(requestLine, Request.class);
	}

	protected void sendResponse(Response response) {
		String responseLine = gsonFormatter.toJson(response);
		System.out.println("sending response " + responseLine);
		synchronized (output) {
			output.println(responseLine);
			output.flush();
		}
	}
}

File: sem4/mpp/project-cs/networking/Worker/AbstractJsonWorker.cs
using Newtonsoft.Json;
using System.Net.Sockets;
using Newtonsoft.Json.Converters;

namespace networking.Worker;

public abstract class AbstractJsonWorker : AbstractWorker {
    protected AbstractJsonWorker(TcpClient connection) : base(connection) {
        JsonConvert.DefaultSettings = () => {
            var settings = new JsonSerializerSettings {
                NullValueHandling = NullValueHandling.Ignore
            };
            settings.Converters.Add(new StringEnumConverter());
            return settings;
        }; 
    }

    protected override Request ReadRequest() {
        try {
            Console.WriteLine("reading request");
            var read = streamReader.ReadLine();
            Console.WriteLine("read " + read);
            return JsonConvert.DeserializeObject<Request>(read);
        }
        catch (Exception e) {
            throw new Exception("Error reading request " + e);
        }
    }

    protected override void SendResponse(Response response) {
        Console.WriteLine("sending response " + response);
        lock (stream) {
            var json = JsonConvert.SerializeObject(response);
            Console.WriteLine("writing: " + json);
            streamWriter.WriteLine(json);
            streamWriter.Flush();
            Console.WriteLine("response sent");
        }
    }
}